compiler/syntax/src/res_core.ml

```diff
@@ -4885,11 +4885,35 @@ and parse_constr_decl_args p =
  *  | constr-name const-args
  *  | attrs constr-name const-args *)
 and parse_type_constructor_declaration_with_bar p =
+  let is_constructor_with_bar p =
+    Parser.lookahead p (fun state ->
+        match state.Parser.token with
+        | DocComment _ -> (
+          Parser.next state;
+          match state.token with
+          | Bar -> true
+          | _ -> false)
+        | Bar -> true
+        | _ -> false)
+  in
   match p.Parser.token with
-  | Bar ->
+  | _ when is_constructor_with_bar p ->
```

Comment:
Nitpick, you are pattern matching where a simple `if/then/else` would do the trick.